File: node.cpp
#include<iostream>
using namespace std;



class Node{
public:
    int data;
    Node* next;

    Node(int d) :data(d), next(nullptr){}
};

Node* Reverse(Node* r){
    Node* curr = r;
    Node* next = nullptr, *pre = nullptr;

    while(curr != nullptr){
        next = curr->next;
        curr->next = pre;
        pre = curr;
        curr = next;
    }
    return pre;
}

void PrintList(Node* r){
    Node* t = r;

    while(t!=nullptr){
        cout<<t->data<<endl;
        t = t->next;
    }
}

int main(){

    Node* r = new Node(1);
    r->next = new  Node(2);
    r->next->next = new Node(3);
    PrintList(r);
    cout<<" ___________________"<<endl;
    PrintList(Reverse(r));
}

File: node.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

func NewNode(v int) *Node {
	var n Node
	n.data = v
	n.next = nil
	return &n
}

func PrintList(n *Node) {
	t := n
	for t != nil {
		fmt.Println(t.data)
		t = t.next
	}
}

func Reverse(n *Node) *Node {
	var curr *Node
	var pre *Node
	var next *Node

	curr = n
	for curr != nil {
		next = curr.next
		curr.next = pre
		pre = curr
		curr = next
	}
	return pre
}

func main() {
	n := NewNode(1)
	n.next = NewNode(2)
	n.next.next = NewNode(3)
	PrintList(n)
	fmt.Println("-----------------")
	PrintList(Reverse(n))
}
